File: solutions/0035_Search_Insert_Position/binary_search.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        if len(nums) == 0: return 0
        if target > nums[-1]: return len(nums)
        return self.bi_search(nums, target, 0, len(nums) - 1)

    def bi_search(self, nums, target, l, r):
        if l == r:
            return l
        m = (l + r) // 2
        if target <= nums[m]:
            return self.bi_search(nums, target, l, m)
        else:
            return self.bi_search(nums, target, m+1, r)

File: solutions/0153_Find_Minimum_in_Rotated_Sorted_Array/binary_search.java
class Solution {
    public int findMin(int[] nums) {
        if (nums.length == 0) return 0;
        int l = 0, r = nums.length - 1;
        while (l < r) {
            int m = (l + r) / 2;
            if (nums[l] > nums[m]) { // pivot in first half
                r = m;
            } else if (nums[r] < nums[m]) { // pivot in second half
                l = m + 1;
            } else { // sorted!
                return nums[l];
            }
        }
        return nums[l];
    }
}
